app/models/projects/acts_as_customizable_patches.rb

```diff
@@ -0,0 +1,248 @@
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Projects::ActsAsCustomizablePatches
+  extend ActiveSupport::Concern
+
+  attr_accessor :_limit_custom_fields_validation_to_section_id, :_query_available_custom_fields_on_global_level
+
+  # attr_accessor :_limit_custom_fields_validation_to_field_id
+  # not needed for now, but might be relevant if we want to have edit dialogs just for one custom field
+
+  included do
+    has_many :project_custom_field_project_mappings, class_name: 'ProjectCustomFieldProjectMapping', foreign_key: :project_id,
+                                                     dependent: :destroy, inverse_of: :project
+    has_many :project_custom_fields, through: :project_custom_field_project_mappings, class_name: 'ProjectCustomField'
+
+    before_save :build_missing_project_custom_field_project_mappings
+    after_save :reset_section_scoped_validation, :set_query_available_custom_fields_to_project_level
+
+    # we need to reset the query_available_custom_fields_on_global_level already after validation
+    # as the update service just calls .valid? and returns if invalid
+    # after_save is not touched in this case which causes the flag to stay active
+    after_validation :set_query_available_custom_fields_to_project_level
+
+    before_update :set_query_available_custom_fields_to_global_level
+
+    before_create :reject_section_scoped_validation_for_creation
+    after_create :disable_custom_fields_with_empty_values
+
+    def build_missing_project_custom_field_project_mappings
+      # activate custom fields for this project (via mapping table) if values have been provided for custom_fields but no mapping exists
+      # current shortcommings:
+      # - boolean custom fields are always activated as a nil value is never provided (always true/false)
+      custom_field_ids = project.custom_values
+        .reject { |cv| cv.value.blank? }
+        .reject { |cv| cv.persisted? } # do not reactivate custom fields which have already been used and disabled
```

Comment:
These two could be made in the same interation:
```suggestion
        .reject { |cv| cv.persisted? || cv.value.blank? } # do not reactivate custom fields which have already been used and disabled
```